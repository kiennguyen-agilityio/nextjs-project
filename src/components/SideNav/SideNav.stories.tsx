import type { Meta, StoryObj } from '@storybook/react';

// components
import SideNav from '@/components/SideNav';

const meta = {
  title: 'Components/SideNav',
  component: SideNav,
} satisfies Meta<typeof SideNav>;

export default meta;

type Story = StoryObj<typeof SideNav>;

export const InputName: Story = {
  args: {},
};
