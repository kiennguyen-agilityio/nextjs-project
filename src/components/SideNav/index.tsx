'use client';

import Link from 'next/link';
import { usePathname } from 'next/navigation';

// constants
import { LINKS } from '@/constants/router';

export default function SideNav() {
  const pathname = usePathname();

  return (
    <div className="flex h-full flex-col px-3 py-4 md:px-2">
      <div className="flex grow flex-row space-x-2 md:flex-col md:space-x-0 md:space-y-2">
        {LINKS.map((link) => {
          const isActive = pathname === link.href;
          const baseClass =
            'flex h-[48px] grow items-center justify-center gap-2 rounded-md p-3 text-sm font-medium hover:bg-sky-100 hover:text-blue-600 md:flex-none md:justify-start md:p-2 md:px-3';
          const activeClass = 'bg-sky-100 text-blue-600';
          const inActiveClass = 'bg-gray-50';
          const linkClasses = `${baseClass} ${isActive ? activeClass : inActiveClass}`;

          return (
            <Link key={link.name} href={link.href} className={linkClasses}>
              <p>{link.name}</p>
            </Link>
          );
        })}
      </div>
    </div>
  );
}
