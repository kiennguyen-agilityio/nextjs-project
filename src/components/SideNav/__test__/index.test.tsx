import { render, screen } from '@testing-library/react';
import { usePathname } from 'next/navigation';

// components
import SideNav from '@/components/SideNav';

jest.mock('next/navigation', () => ({
  usePathname: jest.fn(),
}));

describe('SideNav Component', () => {
  let mockUsePathname: jest.Mock;

  beforeAll(() => {
    mockUsePathname = usePathname as jest.Mock;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders without crashing and matches snapshot', () => {
    const { asFragment } = render(<SideNav />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correct links', () => {
    const { getByText } = render(<SideNav />);

    expect(getByText('Roles')).toBeInTheDocument();
    expect(getByText('Users')).toBeInTheDocument();
  });

  it('applies correct link class when link is active', () => {
    mockUsePathname.mockReturnValue('/roles');
    render(<SideNav />);

    const rolesLink = screen.getByText('Roles');

    expect(rolesLink.parentElement).toHaveClass(
      'flex h-[48px] grow items-center justify-center gap-2 rounded-md bg-sky-100 text-blue-600 p-3 text-sm font-medium hover:bg-sky-100 hover:text-blue-600 md:flex-none md:justify-start md:p-2 md:px-3',
    );
  });

  it('renders Users link as inactive', () => {
    mockUsePathname.mockReturnValue('/roles');
    render(<SideNav />);
    const usersLink = screen.getByText('Users');

    expect(usersLink.parentElement).toHaveClass(
      'flex h-[48px] grow items-center justify-center gap-2 rounded-md bg-gray-50 p-3 text-sm font-medium hover:bg-sky-100 hover:text-blue-600 md:flex-none md:justify-start md:p-2 md:px-3',
    );
  });
});
