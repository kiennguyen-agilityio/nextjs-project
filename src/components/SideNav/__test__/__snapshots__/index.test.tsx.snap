// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNav Component renders without crashing and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex h-full flex-col px-3 py-4 md:px-2"
  >
    <div
      class="flex grow flex-row space-x-2 md:flex-col md:space-x-0 md:space-y-2"
    >
      <a
        class="flex h-[48px] grow items-center justify-center gap-2 rounded-md p-3 text-sm font-medium hover:bg-sky-100 hover:text-blue-600 md:flex-none md:justify-start md:p-2 md:px-3 bg-gray-50"
        href="/roles"
      >
        <p>
          Roles
        </p>
      </a>
      <a
        class="flex h-[48px] grow items-center justify-center gap-2 rounded-md p-3 text-sm font-medium hover:bg-sky-100 hover:text-blue-600 md:flex-none md:justify-start md:p-2 md:px-3 bg-gray-50"
        href="/users"
      >
        <p>
          Users
        </p>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
